enable potatobot to respond with an answer

DavidSnider opened the pull request:
Enable Potatobot to post as instructor's answer. Requires upstream changes to piazza API.

DavidSnider committed:
enable potatobot to respond with an answer

arxanas commented on potatobot/potatobot.py:
```diff
@@ -116,10 +119,21 @@ def _handle_single_post(self, post):
+            for i in self._post_handlers:
+                ret = i(post_info)
+                if ret is not None:
+                    final_answer = "%s<p></p>%s" % (final_answer, i(post_info))
```
Use the `.format` method instead of %-formatting.

Consider separating handlers with a horizontal line of some sort?

This should probably be using `.join` in some way; maybe

```
answers = [i(post_info) for i in self_post_handlers]
final_answer = "<p></p>".join(i for i in answers if i is not None)
```

DavidSnider replied:
I've always used %-formating. pylint even says to do so.

Yeah, join is definitely the answer.

arxanas replied:
Sounds like a bug? http://stackoverflow.com/a/25057746/344643

If you can force `flake8` to flag this then I might budge. But currently all other code in the codebase uses `.format`.

It sounds like [PEP 3101](https://www.python.org/dev/peps/pep-3101/) introduces `format` to replace `%`-formatting.

arxanas commented on potatobot/potatobot.py:
```diff
@@ -116,10 +119,21 @@ def _handle_single_post(self, post):
+                ret = i(post_info)
+                if ret is not None:
+                    final_answer = "%s<p></p>%s" % (final_answer, i(post_info))
+            if final_answer and not [d for d in post[
```
More idiomatic is `any`, i.e.

```
if final_answer and not any(d for d in post['change_log'] if d['type'] == 'i_answer'):
    ...
```

I'm pretty sure all other strings in the codebase use double-quotes, so do that.

arxanas commented on potatobot/potatobot.py:
```diff
@@ -116,10 +119,21 @@ def _handle_single_post(self, post):
+                    final_answer = "%s<p></p>%s" % (final_answer, i(post_info))
+            if final_answer and not [d for d in post[
+                    'change_log'] if d['type'] == 'i_answer']:
+                self._network.create_answer(post, final_answer, 0)
```
What does `0` mean? Use a keyword argument to elucidate.

arxanas commented on potatobot/potatobot.py:
```diff
@@ -143,6 +157,9 @@ def _should_ignore_post(self, post):
                                  post["nr"]
                                  ))
                 return True
+        for i in post['change_log']:
```
If you need to do this answer check in more than one place (above and here) then it should be in a helper function.

arxanas commented on profile281.py:
```diff
@@ -209,11 +209,11 @@ def learn_to_valgrind_please(post_info):
     def check_for_duplicate_posts(post_info):
         if post_info.status != "private":
-            
+
```
:+1:

arxanas commented:
Link to `piazza_api` PR?

arxanas commented on potatobot/potatobot.py:
```diff
@@ -116,10 +119,21 @@ def _handle_single_post(self, post):
+            for i in self._post_handlers:
+                ret = i(post_info)
+                if ret is not None:
+                    self._network.create_followup(post, ret)
```
Considering that both branches need to get non-`None` responses in order, that should probably be done first and abstracted out, i.e.

```
responses = [i(post_info) for i in post_handlers)]
responses = [i for i in responses if i is not None]

if self._post_as_answer:
    ...
else:
    ...
```

DavidSnider commented:
requires hfaran/piazza-api#31

DavidSnider committed:
make requested changes

DavidSnider committed:
change to keep in line with Piazza API

DavidSnider committed:
change to keep in line with Piazza API

arxanas commented on potatobot/potatobot.py:
```diff
@@ -119,21 +119,18 @@ def _handle_single_post(self, post):
-            if final_answer and not [d for d in post[
-                    'change_log'] if d['type'] == 'i_answer']:
-                self._network.create_answer(post, final_answer, 0)
+            if not self.has_answer(post):
```
Hasn't this already been checked in `_should_ignore_post`?

DavidSnider replied:
this is accurate

DavidSnider committed:
Merge branch 'RepondAsAnswer' of github.com:DavidSnider/potatobot into RepondAsAnswer

Conflicts:
	potatobot/potatobot.py

arxanas committed:
Handlers can choose whether they want to post an answer or a follow-up

Usage:

    from potatobot import Answer

    @bot.handle_post
    def handle(post_info):
        return Answer("this is an answer")

DavidSnider committed:
better message for result similarity

arxanas committed:
Switch to answers for 281 post handlers

arxanas committed:
Merge branch 'RepondAsAnswer' of https://github.com/DavidSnider/potatobot into return-response

DavidSnider committed:
Merge pull request #1 from arxanas/return-response

Handlers can choose whether they want to post an answer or a follow-up

DavidSnider committed:
added spaces and dashes between posts in answer

arxanas commented:
If you update requirements.txt we should be good to go.

DavidSnider committed:
updates requirements

DavidSnider commented:
gogogo
